vim: Add Separator and RemoveIndent in Join Lines, fix gJ use space join

0x2CA opened the pull request:
Closes #22492

Release Notes:

- Added Join Lines Separator And RemoveIndent

0x2CA committed:
separator

0x2CA committed:
fix clippy

0x2CA committed:
add remove_indent

0x2CA committed:
fix

ConradIrwin commented:
@0x2CA Thank you for this!

Using a `struct JoinLines { ... }` causes the action to not show up in the command palette. We should keep `editor::JoinLines` as is.

For vim it matters a bit less, so we can change that one; or we could add `vim::JoinLinesNoSpace`. 

Is the `removeIndent` option needed for compatibility, or just something you thought might be useful?

Could you please also add a test for `gJ`:

```
    #[gpui::test]
    async fn test_gj(cx: &mut gpui::TestAppContext) {
        let mut cx = NeovimBackedTestContext::new(cx).await;

        cx.set_shared_state(indoc! {"
            ˇa
            b
        })
        .await;
        cx.simulate_shared_keystrokes("g shift-j").await;
        cx.shared_state().await.assert_eq(indoc! {"
            aˇb
        })
    }
```

You can run the test with `cargo test -p vim --features neovim test_gj` and commit the JSON file it generates (so that the CI machines don't need to run real neovim)

0x2CA committed:
fix editor command

0x2CA committed:
add g shift-j test

0x2CA commented:
@ConradIrwin 

because `gJ` not only separates character differences, but does not remove or insert spaces.

```
gJ			Join [count] lines, with a minimum of two lines.
			Don't insert or remove any spaces.
```

for example

```
a
    b
```


```
a    b
```

I have fixed the command and added the test.

ConradIrwin committed:
Simplify implementation

ConradIrwin commented:
@0x2CA Thanks for this. Thinking about it more, it's probably easiest to keep the implementation less complex (although it's less flexible). If you'd like to add a `vim::JoinLinesRemoveWhitespace` in addition for that use-case Id' be ok with that.

0x2CA commented:
Removing the indentation is also an essential option, which will result in the removal of spaces, and my implementation is freer to use any hyphenation, allowing the user to implement hyphens like "a_b_c_d"
